lib/AST/TS2Flow.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "hermes/AST/TS2Flow.h"
+
+#include "hermes/AST/RecursiveVisitor.h"
+
+namespace hermes {
+
+namespace {
+
+class TS2FlowValidator {
+ public:
+  explicit TS2FlowValidator(Context &context) : context_(context) {}
+
+  /// Check if the given program has any TS nodes.
+  /// \param program the top-level program node.
+  /// \return true if the program has no TS node, false otherwise.
+  bool validate(ESTree::ProgramNode *program) {
+    auto &sm = context_.getSourceErrorManager();
+    if (sm.getErrorCount())
+      return false;
+
+    visitESTreeNode(*this, program);
+
+    return sm.getErrorCount() == 0;
+  }
+
+  bool incRecursionDepth(ESTree::Node *) {
+    return true;
+  }
+
+  void decRecursionDepth() {}
+
+  void visit(ESTree::Node *node) {
+    auto kind = node->getKind();
+
+    // TSTypeAnnotation is the first TS node and TSModifiers is the last.
+    // This must be in sync with ESTree.def.
+    if (kind >= ESTree::NodeKind::TSTypeAnnotation &&
+        kind <= ESTree::NodeKind::TSModifiers) {
+      return reportError(node);
+    }
+
+    // Update/restore parent_ before/after visiting the children,
+    // so we can keep track of parent_ when we cannot define
+    // `void visit(ESTree::Node *node, ESTree::Node *parent)`.
```

Comment:
Why can't we use this two-argument overload of `visit`? We use it in other places and `RecursiveVisitor` should do all the parent passing for you.

Reply:
To my understanding, `void visit(ESTree::Node *node, ESTree::Node *parent)` (the base case with parent) will never get called because the type of `node` is narrowed by the visitor. Or, am I missing something?

Reply:
Here is a commit by @avp that fixes this. We will land it tomorrow but meanwhile you can rebase on top of it and see whether it addresses the problem.

Reply:
https://github.com/facebook/hermes/pull/1108

Reply:
Landed: https://github.com/facebook/hermes/commit/c667423d16c156cecb3b85912955b00cda13abd2

Reply:
Thank you! I've just rebased and pushed the revision. Could you please take a look?